### PR TITLE
CR-1125075 xbutil dd read incorrectly reports success despite a DMA failure

### PR DESCRIPTION
**Pull request by dbenusov-xilinx**
#### Problem solved by the commit
During an xbutil advanced --read-mem if an error is raised while reading, the xbutil command will return successfully without returning an error.

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
Discovered via ASTeR testing.

#### How problem was solved, alternative solutions (if any) and why they were rejected
The problem was solved by returning the error code generated by the bo migration function.

#### Risks (if any) associated the changes in the commit
None.

#### What has been tested and how, request additional testing if necessary
Tested on a U30 on Ubuntu 20.04

Failure Test
The test had the driver modified to return a failure to simulate the problem seen in the CR.
User side
```
dbenusov@xsjsagarw50:/proj/rdi/staff/dbenusov$ xbutil advanced --read-mem --address 0x800000000 --size 1024 -o /tmp/test.txt --force -d 17:00
ERROR: perform_memory_action: Invalid argument - reading 1024 bytes from DDR[0](0x800000000)
dbenusov@xsjsagarw50:/proj/rdi/staff/dbenusov$ echo $?
1
```
Associated Dmesg log
```
[  +0.000024] xocl 0000:17:00.1:  ffff9becc7f7e0b0 xocl_create_client: created KDS client for pid(40258), ret: 0
[  +0.018072] xocl 0000:17:00.1: icap.u.22020096 ffff9bec2bb5a010 icap_cached_ocl_frequency: no cached data for 3
[  +0.001233] xocl 0000:17:00.1: icap.u.22020096 ffff9bec2bb5a010 icap_lock_bitstream: bitstream c6ec5221-0e45-4707-b64e-2fd7da107abb locked, ref=1
[  +0.000007] xocl 0000:17:00.1:  ffff9becc7f7e0b0 kds_add_context: Client pid(40258) add context Domain(65535) CU(0xffff) shared(true)
[  +0.001534] xocl 0000:17:00.1: dma.xdma.u.5242880 ffff9be9c11a9010 xdma_migrate_bo: DMA failed, Dumping SG Page Table, ep addr 800000000
[  +0.000003] xocl 0000:17:00.1: dma.xdma.u.5242880 ffff9be9c11a9010 xdma_migrate_bo: 0, 0xa51015000
[  +0.000002] Migrated dma status: -22
[  +0.000001] Migrated dma rest: 4294967274
[  +0.000001] Migrate BO status: -22
```

Regular case
User side
```
dbenusov@xsjsagarw50:/proj/rdi/staff/dbenusov$ xbutil advanced --read-mem --address 0x800000000 --size 1024 -o /tmp/test.txt --force -d 17:00
Memory read succeeded
dbenusov@xsjsagarw50:/proj/rdi/staff/dbenusov$ echo $?
0
```
Associated Dmesg log
```
[  +0.000019] xocl 0000:65:00.1:  ffff9becc7fef0b0 xocl_create_client: created KDS client for pid(48614), ret: 0
[  +0.002667] xocl 0000:65:00.1: xmc.u.18874370 ffff9becbecb6410 xmc_read_from_peer: reading from peer
[  +0.000024] xocl 0000:65:00.1: mailbox.u.9437186 ffff9becbecb0810 _mailbox_request: sending request: 10 via HW
[  +0.101540] xclmgmt 0000:65:00.0: mailbox.m.9437186 ffff9beb249e1410 process_request: received request from peer: 10, passed on
[  +0.000012] xclmgmt 0000:65:00.0: xclmgmt_read_subdev_req: req kind 0
[  +0.000018] xclmgmt 0000:65:00.0: mailbox.m.9437186 ffff9beb249e1410 mailbox_post_response: posting response for: 10 via HW
[  +0.028447] xocl 0000:65:00.1: icap.u.22020098 ffff9becbecb6010 icap_read_from_peer: reading from peer
[  +0.000010] xocl 0000:65:00.1: mailbox.u.9437186 ffff9becbecb0810 _mailbox_request: sending request: 10 via HW
[  +0.000317] xclmgmt 0000:65:00.0: mailbox.m.9437186 ffff9beb249e1410 process_request: received request from peer: 10, passed on
[  +0.000012] xclmgmt 0000:65:00.0: xclmgmt_read_subdev_req: req kind 1
[  +0.561062] xclmgmt 0000:65:00.0: mailbox.m.9437186 ffff9beb249e1410 mailbox_post_response: posting response for: 10 via HW
[  +0.000799] xocl 0000:65:00.1: icap.u.22020098 ffff9becbecb6010 icap_cached_ocl_frequency: no cached data for 3
[  +0.073091] xocl 0000:65:00.1:  ffff9becc7fef0b0 xocl_destroy_client: client exits pid(48614)
[  +0.000013] xocl 0000:65:00.1:  ffff9becc7fef0b0 xocl_drvinst_close: CLOSE 2
```

#### Documentation impact (if any)
None

### CODE REVIEW

**Review by maxzhen — COMMENTED** [review]
There are other cases where xocl_migrate_bo() is called in the same file. Can we do the same here, which seems to be simpler?

**Comment by dayeh-xilinx** [comment]
retest this please. Jenkins server restart.

**Comment by dbenusov-xilinx** [comment]
> There are other cases where xocl_migrate_bo() is called in the same file. Can we do the same here, which seems to be simpler?

Good plan. I ended up moving the -EIO check into the migrate functions in xdma and qdma rather than having the higher level functions need to check. We are not able to see how many bytes are processed, but, no one was using those values anyways.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xdma.c [hunk 1]
```diff
@@ -66,19 +66,28 @@ static ssize_t xdma_migrate_bo(struct platform_device *pdev,
...
 
 	xdma = platform_get_drvdata(pdev);
 	xocl_dbg(&pdev->dev, "TID %d, Channel:%d, Offset: 0x%llx, Dir: %d",
 		pid, channel, paddr, dir);
-	ret = xdma_xfer_fastpath(xdma->dma_handle, channel, dir,
+	migrated = xdma_xfer_fastpath(xdma->dma_handle, channel, dir,
 		paddr, sgt, false, 10000);
-	if (ret >= 0) {
+	if (migrated >= 0) {
 		xdma->channel_usage[dir][channel] += ret;
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/xdma.c:79]: I would not recommend making this type of change so close to a release. Or maybe you could make sure you can test qdma and xdma carefully to make sure you did not introduce any regression. E.g., here you probably failed to update channel_usage correctly.

File: src/runtime_src/core/common/memaccess.cpp [hunk 2]
```diff
@@ -165,12 +167,24 @@ perform_memory_action(xrt_core::device* device, xrt_core::aligned_ptr_type& buf,
 
     // Update the buffer index based on how far we have read
     void* current_buffer_location = static_cast<char *>(buf.get()) + bytes_seen;
+    boost::format err_fmt("%s: Code : %d - %s %u bytes from %s(0x%x)");
+    err_fmt % __func__;
```
Comment (rozumx) [inline src/runtime_src/core/common/memaccess.cpp:171]: Hmm...does this like of code doing anything?

Looking at the code I see you are expecting that the first argument is expecting the name of the function.   Have you tested it to see if it actuals prints the expected message.

If it does work, then I learned something new today and would like to understand more on how this works.

File: src/runtime_src/core/common/memaccess.cpp [hunk 3]
```diff
@@ -165,12 +167,24 @@ perform_memory_action(xrt_core::device* device, xrt_core::aligned_ptr_type& buf,
...
     switch (action) {
       case operation_type::read:
-        device->unmgd_pread(current_buffer_location, bytes_to_edit, validated_start_addr);
+        try {
+          device->unmgd_pread(current_buffer_location, bytes_to_edit, validated_start_addr);
+        } catch (const std::exception& e) {
+          auto err_msg = err_fmt % errno % "reading" % bytes_to_edit % it->m_tag % current_addr;
```
Comment (rozumx) [inline src/runtime_src/core/common/memaccess.cpp:177]:
```suggestion
         const auto err_msg = err_fmt % errno % "reading" % bytes_to_edit % it->m_tag % current_addr;
```

File: src/runtime_src/core/common/memaccess.cpp [hunk 4]
```diff
@@ -165,12 +167,24 @@ perform_memory_action(xrt_core::device* device, xrt_core::aligned_ptr_type& buf,
...
+          throw xrt_core::error(std::errc::operation_canceled, err_msg.str());
+        }
         break;
       case operation_type::write:
-        device->unmgd_pwrite(current_buffer_location, bytes_to_edit, validated_start_addr);
+        try {
+          device->unmgd_pwrite(current_buffer_location, bytes_to_edit, validated_start_addr);
+        } catch (const std::exception& e) {
+          auto err_msg = err_fmt % errno % "writing" % bytes_to_edit % it->m_tag % current_addr;
```
Comment (rozumx) [inline src/runtime_src/core/common/memaccess.cpp:185]:
```suggestion
          const auto err_msg = err_fmt % errno % "writing" % bytes_to_edit % it->m_tag % current_addr;
```

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_bo.c [hunk 5]
```diff
@@ -919,12 +920,11 @@ static int xocl_migrate_unmgd(struct xocl_dev *xdev, uint64_t data_ptr, uint64_t
 {
 	int channel;
 	struct drm_xocl_unmgd unmgd;
-	int ret;
-	ssize_t migrated;
+	ssize_t ret;
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_bo.c:923]: Uninitialized variables.  Please initialize ALL variables in the function.

**Review by rozumx — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/common/memaccess.cpp (hunk 2 above)
Comment (dbenusov-xilinx) [reply src/runtime_src/core/common/memaccess.cpp:171]: It does work! I found it to be a neat trick when the message being generated used the same string. One issue I have is (as far as I know) the parameters need to be placed in order.
```
dbenusov@xsjsagarw50:/proj/rdi/staff/dbenusov$ xbutil advanced --read-mem --address 0x800000000 --size 1024 -o /tmp/test.txt --force -d 17:00
ERROR: perform_memory_action: Invalid argument - reading 1024 bytes from DDR[0](0x800000000)
```

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Review by maxzhen — APPROVED** [review]

**Review by chienwei-lan — APPROVED** [review]

File: src/runtime_src/core/common/memaccess.cpp (hunk 2 above)
Comment (rozumx) [reply src/runtime_src/core/common/memaccess.cpp:171]: Time for me to be educated with regards to this syntax.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_bo.c [hunk 6]
```diff
@@ -917,14 +918,13 @@ int xocl_info_bo_ioctl(struct drm_device *dev,
 
 static int xocl_migrate_unmgd(struct xocl_dev *xdev, uint64_t data_ptr, uint64_t paddr, size_t size, bool dir)
 {
-	int channel;
-	struct drm_xocl_unmgd unmgd;
-	int ret;
-	ssize_t migrated;
+	int channel = 0;
+	struct drm_xocl_unmgd unmgd = {0};
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_bo.c:922]: zero (0) not needed.

```suggestion
	struct drm_xocl_unmgd unmgd = {};
```

**Review by rozumx — APPROVED** [review]
Approve with a minor initialization request.

**Pull request merged**